No lattice repl

brooksmtownsend opened the pull request:
Addresses #92 

No-nats mode now supports all but the following commands from the standard lattice connected mode:
1. `ctl get claims`
2. `ctl update actor`
3. `ctl get inventory` (partial)
These are all unsupported at the moment because there are not equivalent commands to invoke on the host directly, and some make use of the host controller on the control interface side in order to return the correct information. This will need to be addressed. I'm primarily concerned about implementing `get inventory` for the host inventory, as that contains much needed information, specifically link names and OCI references. @autodidaddict I've added what I can of support for now (see the host inventory implementation for standalone) however we might be better off implementing some of this functionality on the host side.

For this PR, I've added a few additional threads and channels to allow `wash up` to run a REPL without connecting to NATS, as indicated by the `Standalone` title by the REPL. This launches an embedded host, as usual, however, commands will be invoked directly on that host instead of proxying through the control interface.

A few modifications for reviewers:
- `ctl` and `up` are getting to contain large amounts of code, so I've split some large pieces into separate files under a folder of the same name. This is a pretty standard Rust pattern. I didn't split larger pieces of the REPL out into separate files because I made many assumptions about accessing fields, and I'd have to write just as much code as I'd save in getters, setters, and modifications.
- the `output` module under `ctl/output.rs` is used to standardize the output format for control interface operations and embedded host operations, that way what the user sees should be identical for standalone mode and lattice connected mode.
- I edited some integration tests to better check for failures while I was refactoring the output modules.

autodidaddict commented:
Safe? You think the weekend makes you safe?
![image](https://user-images.githubusercontent.com/13748641/110176923-168ee600-7dd2-11eb-8945-85b8018346d3.png)

brooksmtownsend commented:
@autodidaddict Yeah it's just moved around because I restructured the project a bit to clean up the code, check `src/up/mod.rs` https://github.com/wasmCloud/wash/pull/107/files#diff-dcf780801376c219d3b5d7f2f9a5a2745c5ccb124b8c621281e13d3c775ed9f3

autodidaddict commented:
Thanks for the link.. though it's still not in the diff, this is what I was looking for:

https://github.com/wasmCloud/wash/blob/dd0540b3e56891689a10ac8274591948794c8920/src/up/mod.rs#L642

brooksmtownsend commented:
As of my latest commit:
- Fixed bug where characters would show vertically across output window
- Implemented the above three functions, which now brings standalone mode to full parity with lattice connected mode

There are only a few more areas where I need to implement functionality `wasmcloud-host`:
- [ ]  Need a way to get the provider link name for host inventory
- [ ]  Need a way to update the OCI mapping when an actor is updated (currently if I update `echo:0.2.0` to `echo:0.2.1`, the OCI mapping still contains the echo public key and `echo:0.2.0`)
- [ ]  Need to have connection parameters provided for update actor, currently they are just defaults (no latest tags, secure connection, no user/pass)

brooksmtownsend commented:
All functionality has been implemented and `wash` no-nats (standalone) mode has exact parity with nats (lattice) mode. The build is going to fail because of relative path dependencies and that can be remedied once https://github.com/wasmCloud/wasmCloud/pull/163 is ready

brooksmtownsend commented:
REVIEWER'S GUIDE TO SOME OF THE BIG CHANGES (@autodidaddict )

I've refactored the `ctl` and `up` modules so they appear to have been deleted and written. The `ctl` module has been refactored into `mod` and `output`, where `mod` is the functionality of the `ctl` module and `output` is where all of the helper functions exist to translate the CTL resources into a String for output in the terminal. The neat thing about this is that I'm using these helper functions with the standalone mode as well, that way there will be no difference between what the user sees in either mode.

`up` was refactored as well, mostly to pull some of the large parts of standalone mode types into another file. I'd like to continue to refactor this a bit as `up` and `ctl` are approaching 1000 lines but for now this is enough change.

autodidaddict reviewed (changes requested):
LGTM once clippy has been placated.
